### PR TITLE
Feature/panda 3214 my coupons

### PR DESCRIPTION
**Pull request by Terry01Tan**
Internal review for my coupons.

### CODE REVIEW

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    ]).pipe(
+      map(([subscribing, unsubscribing]) => subscribing || unsubscribing)
+    );
+  }
+
+  getSortLabels(): Observable<{
+    byStartDateAsc: string;
+    byStartDateDesc: string;
+    byEndDateAsc: string;
+    byEndDateDesc: string;
```
Comment (Jerify02) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts]: This should be a property, if it's used in a template. Function will be recalculaten in each change detection cycle, and there will be new observable created each time.

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+      .getCustomerCoupons(this.PAGE_SIZE)
+      .pipe(
+        tap(
+          coupons =>
+            (this.pagination = {
+              currentPage: coupons.pagination.page,
+              pageSize: coupons.pagination.count,
+              totalPages: coupons.pagination.totalPages,
+              totalResults: coupons.pagination.totalCount,
+              sort: this.sort,
```
Comment (Jerify02) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts]: Maybe we should do this converting in a normalizer, eg: `core/src/occ/adapters/cart/converters/occ-cart-normalizer.ts`.

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
```
Comment (Jerify02) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.ts]: 'OnInit' is declared but its value is never read.

File: projects/core/src/user/services/find-product-page-meta.resolver.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+      map(([title, breadcrumbs]: [string, any[]]) => ({
+        title,
+        breadcrumbs,
+      }))
+    );
+  }
+
+  resolveBreadcrumbs(): Observable<any[]> {
+    const breadcrumbs = [];
+    breadcrumbs.push({ label: 'Home', link: '/' });
```
Comment (Seven7hang) [inline projects/core/src/user/services/find-product-page-meta.resolver.ts]: `const breadcrumbs = [{ label: 'Home', link: '/' }];`

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.html [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+      <div class="cx-coupon-card-row bottom">
+        <div class="cx-coupon-notification form-check">
+          <label>
+            <input
+              type="checkbox"
+              class="form-check-input"
+              [checked]="coupon?.notificationOn"
+              [class.disabled]="couponSubscriptionLoading$ | async"
+              [disabled]="couponSubscriptionLoading$ | async"
+              (change)="notificationChange()"
```
Comment (Seven7hang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.html]: suggest using: onSubscriptionChange()

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.html [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            />
+            <span class="form-check-label">
+              {{ 'myCoupons.notification' | cxTranslate }}
+            </span>
+          </label>
+        </div>
+
+        <div class="cx-coupon-find-product col-lg-6 col-md-12 col-sm-6">
+          <button class="btn btn-block btn-action" (click)="findProduct()">
```
Comment (Seven7hang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.html]: `(click)="findProducts()"`

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    this.modalRef = this.modalService.open(CouponDialogComponent, {
+      centered: true,
+      size: 'lg',
+    });
+
+    modalInstance = this.modalRef.componentInstance;
+    modalInstance.coupon = this.coupon;
+  }
+
+  findProduct(): void {
```
Comment (Seven7hang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/coupon-card/coupon-card.component.ts]: `findProducts()`

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.html [hunk 8]
```diff
@@ -0,0 +1,82 @@
+<div class="cx-section">
+  <ng-container *ngIf="!(couponsLoading$ | async); else loading">
+    <ng-container *ngIf="couponResult$ | async as couponResult">
+      <div class="cx-my-coupons-header">
+        <h3>{{ 'myCoupons.myCoupons' | cxTranslate }}</h3>
+      </div>
+
+      <ng-container
+        *ngIf="couponResult.pagination.totalCount > 0; else noCoupons"
```
Comment (Seven7hang) [inline projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.html]: check if this judgement is duplicated with line#3

File: projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts (hunk 1 above)
Comment (Terry01Tan) [reply projects/storefrontlib/src/cms-components/myaccount/my-coupons/my-coupons.component.ts]: Resolved.

File: projects/core/src/occ/occ-models/occ-endpoints.model.ts [hunk 9]
```diff
@@ -275,4 +275,10 @@ export interface OccEndpoints {
    * @member {string}
    */
   cartVoucher?: string;
+  /**
+   * Endpoint for coupons
+   *
+   * @member {string}
+   */
+  myCoupons?: string;
```
[the remark this answers is not in the document]
Comment (Terry01Tan) [reply projects/core/src/occ/occ-models/occ-endpoints.model.ts]: Resolved

File: projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.spec.ts [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+import {
+  CustomerCoupon,
+  CustomerCouponSearchResult,
+  CustomerCouponNotification,
+  CustomerCoupon2Customer,
+} from '../../../model/customer-coupon.model';
+import { OccCustomerCouponAdapter } from './occ-customer-coupon.adapter';
+import { MockOccEndpointsService } from './unit-test.helper';
+import { OccEndpointsService } from '../../services/occ-endpoints.service';
+import { ConverterService } from 'projects/core/src/util/converter.service';
```
Comment (Qingyu-Yang) [inline projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.spec.ts]: Use ../..

File: projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.spec.ts [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+      baseUrl: '',
+      prefix: '',
+    },
+  },
+};
+
+describe('OccCustomerCouponAdapter', () => {
+  let occCustomerCouponAdapter: OccCustomerCouponAdapter;
+  let httpMock: HttpTestingController;
+  let converter: ConverterService;
```
Comment (Qingyu-Yang) [inline projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.spec.ts]: need to add test case `should use convertor`
refer to https://github.com/SAP/cloud-commerce-spartacus-storefront/pull/4137/files#diff-3ce24e75b5602b029869a48d5548952cR253

File: projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    userId: string,
+    couponCode: string
+  ): Observable<CustomerCouponNotification> {
+    const url = this.occEndpoints.getUrl('couponNotification', {
+      userId,
+      couponCode,
+    });
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
+      'Cache-Control': 'no-cache',
```
Comment (Qingyu-Yang) [inline projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.ts]: what dose `'Cache-Control': 'no-cache',` mean? If it is unnecessary, you can extra a method called protect getHeaders() and call it in each method

File: projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    const headers = new HttpHeaders({
+      'Content-Type': 'application/json',
+    });
+
+    return this.http
+      .get<Occ.CustomerCouponSearchResult>(url, { headers, params })
+      .pipe(
+        catchError((error: any) => throwError(error)),
```
Comment (Qingyu-Yang) [inline projects/core/src/occ/adapters/user/occ-customer-coupon.adapter.ts]: In the latest develop branch, `catchError` and `throwError` is not used

File: projects/core/src/user/store/effects/customer-coupon.effect.spec.ts [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+  CustomerCouponSearchResult,
+  CustomerCouponNotification,
+  CustomerCoupon2Customer,
+} from '../../../model/customer-coupon.model';
+import { User } from '../../../model/misc.model';
+import { CustomerCouponConnector } from '../../connectors/customer-coupon/customer-coupon.connector';
+import { CustomerCouponAdapter } from '../../connectors/customer-coupon/customer-coupon.adapter';
+import { UserService } from '../../facade/user.service';
+import * as fromCustomerCouponsAction from '../actions/customer-coupon.action';
+import * as fromCustomerCouponsEffect from './customer-coupon.effect';
```
Comment (Qingyu-Yang) [inline projects/core/src/user/store/effects/customer-coupon.effect.spec.ts]: Current usage is 
```
import { XXXAction } from '../actions/xxx.action';
import { XXXEffect } from './xxx.effect';
```

File: projects/core/src/user/store/effects/customer-coupon.effect.ts [hunk 15]
```diff
@@ -0,0 +1,135 @@
+import { Injectable } from '@angular/core';
+import { Actions, Effect, ofType } from '@ngrx/effects';
+import { Observable, of } from 'rxjs';
+import { catchError, map, mergeMap } from 'rxjs/operators';
+import * as fromCustomerCouponsAction from '../actions/customer-coupon.action';
```
Comment (Qingyu-Yang) [inline projects/core/src/user/store/effects/customer-coupon.effect.ts]: import {CustomerCouponsActions} from '../actions/customer-coupon.action'

File: projects/storefrontapp-e2e-cypress/cypress/helpers/my-coupons.ts [hunk 16]
```diff
@@ -0,0 +1,236 @@
+import { generateMail, randomString } from './user';
+import { login } from './auth-forms';
+import { standardUser } from '../sample-data/shared-users';
+
+// export const configurableProductCode = '1934793';
```
Comment (Qingyu-Yang) [inline projects/storefrontapp-e2e-cypress/cypress/helpers/my-coupons.ts]: remove comments

Comment (Qingyu-Yang) [reply projects/storefrontapp-e2e-cypress/cypress/helpers/my-coupons.ts]: and why `storefrontapp-e2e-cypress/cypress/integration/regression/my-account/my-coupons.e2e-spec.ts` is a empty file?

**Pull request closed without merging**